Renderers/src/artofillusion/raytracer/RTCube.java

```diff
@@ -128,6 +129,8 @@ public SurfaceIntersection checkIntersection(Ray r)
   {
     Vec3 rorig = r.getOrigin(), rdir = r.getDirection();
     Vec3 origin, direction;
+    double cubeTol = Math.max(Math.max(Math.max(maxx-minx, maxy-miny), maxz-minz), rorig.length()*0.01)*TOL;
```

Comment:
I thought, when we first discussed this, that this could be an instance variable that was calculated when the `RTCube` was constructed.

When did it acquire the dependency on the length of the Ray origin?

Reply:
> I thought, ... that this could be an instance variable

So did I at the time, but it has been like this for a while already. 

The original idea works as long as you stay relatively close to the world origin but when the objects travel further away the light ray problem returns. It happens independently in all coordinate directions. Some of the test scenes I posted demonstrate just the phenomenon, that both the size and the distance from origin have a part here independent of each others.

But when I now started to test a bit more, it turned out that at least in the case of `RTCube` I could drop the object size from the equation and use only the ray origin. For the more complex objects it seems a bit more complicated....

What is behind it is that numerical error with double precision numbers is usually around 1E-16 but that only applies to numbers close to 1.0. The error naturally increases while the number increases so for 1E5 the error is around 1E-11 and the ray origin is defined in world coordinates.

**So back to square one.** I think I'll  reset this branch and rethink.

For one thing, I don't like myself that the object surface tolerance is recalculated for every ray. I'm not too crazy about calculating the square root in the `.length()` either but that worked. On the other hand, if it was going to be done this way there was no reason to keep the object variable in any more. _(At some point I even had the instance variable and a couple of temporary variables before the final decision about the tolerance, but I compressed those out of the code.)_

Looking at it now, the idea that might work could be to find the point of the object (well, it's bounding box actually) that has the greatest distance from the world origin and select the max absolute value of the coordinates as the base of the tolerance. This tolerance would be used as the instance variable.

Then there is the problem with round surfaces of spheres and cylinders. The tolerance is compared to values calculated from `r2` so the numerical error increases (somehow) relative to the 2nd power of the size of the object while it increases in the 1st power of distance. I did some initial testing with `Cylinder`. What I got so far was not yet an universal solution but it seemed like a possible track.

Anyway the 1E-12 tolerance is extremely ambitious to begin with. In real world dimensions it would mean one nanometer for one kilometer (!). I doubt if anybody would ever render an image that requires that good a separation (well, I might, but...). We probably could afford to loosen it a couple of decimal positions if it comes to that.